net/multi_listen.go

```diff
@@ -152,29 +151,27 @@ func (ml *multiListener) Accept() (net.Conn, error) {
 // the go-routines to exit.
 func (ml *multiListener) Close() error {
 	// Make sure this can be called repeatedly without explosions.
-	if !ml.closed.CompareAndSwap(false, true) {
-		return fmt.Errorf("use of closed network connection")
-	}
-
-	// Tell all sub-listeners to stop.
-	close(ml.stopCh)
-
-	// Closing the listeners causes Accept() to immediately return an error in
-	// the sub-listener go-routines.
-	for _, l := range ml.listeners {
-		_ = l.Close()
-	}
+	ml.closeOnce.Do(func() {
+		// Tell all sub-listeners to stop.
+		close(ml.stopCh)
+
+		// Closing the listeners causes Accept() to immediately return an error in
+		// the sub-listener go-routines.
+		for _, l := range ml.listeners {
+			_ = l.Close()
+		}
 
-	// Wait for all the sub-listener go-routines to exit.
-	ml.wg.Wait()
-	close(ml.connCh)
+		// Wait for all the sub-listener go-routines to exit.
+		ml.wg.Wait()
+		close(ml.connCh)
 
-	// Drain any already-queued connections.
-	for connErr := range ml.connCh {
-		if connErr.conn != nil {
-			_ = connErr.conn.Close()
+		// Drain any already-queued connections.
+		for connErr := range ml.connCh {
+			if connErr.conn != nil {
+				_ = connErr.conn.Close()
+			}
 		}
-	}
+	})
 	return nil
```

Comment:
I think it doesn't make sense to return an error for double close. Either we return the actual underlying errors or no error at all. Any reason to create "artificial" errors? I.e. nothing bad happened because of the second call, why return an error?